### DISCUSSION TITLE
Record MIDI events

### DISCUSSION THREAD

**Discussion by kevintomsgithub** [mido/mido#285] [Ideas]
Hey guys,

Lately I was bulding a MIDI recorded and came over the requirement of recording the note events. So I was giving it a thought to include a MIDI recorder class which will have a record flag. All the note event, after setting the record flag to true will be recorded until it is again set to false.

What are your thoughts on the same?

Thanks.

**Comment by olemb** [comment]
I think this is a great idea.

The current `MidiFile` does too many things already. Ideally I'd like to change the API so there's a `read_midi_file()` which returns the MIDI data and `write_midi_file()` to save it. This will make it easy write `MidiPlayer` and `MidiRecorder` or whatever one wants.

`python-midi` does this and it's a much cleaner API: https://github.com/vishnubob/python-midi

Not sure about the name `MidiPattern` though. Maybe `MidiData`, or is that too generic?

**Reply by arturbecker** [reply]
Hey @olemb!
I'm trying to record incoming MIDI events, so that I can play them back afterwards. From what I get in this discussion, there's no simple way to to it in Mido right now, right?
For my use case I'm thinking of implementing a logic such as:

1. Receive MIDI message (e.g. note_on, value=64)
2. Wait
3. Receive MIDI message, note_off, of the same value as a note currently open (e.g. 64)
4. Convert the "ticks" between note_on and note_off of note 64 to a duration in the note_on message

Do you think that makes sense?
